src/jackdaw/streams/xform.clj

```diff
@@ -0,0 +1,67 @@
+(ns jackdaw.streams.xform
+  "Helper functions for working with transducers."
+  (:gen-class)
+  (:require [jackdaw.serdes :as js]
+            [jackdaw.streams :as j])
+  (:import org.apache.kafka.streams.kstream.Transformer
+           org.apache.kafka.streams.KeyValue
+           [org.apache.kafka.streams.state KeyValueStore Stores]
+           org.apache.kafka.streams.StreamsBuilder))
+
+(defn fake-kv-store
+  "Creates an instance of org.apache.kafka.streams.state.KeyValueStore
+  with overrides for get and put."
+  [init]
+  (let [store (volatile! init)]
+    (reify KeyValueStore
+      (get [_ k]
+        (clojure.core/get @store k))
+
+      (put [_ k v]
+        (vswap! store assoc k v)))))
+
+(defn kv-store-swap-fn
+  "Takes an instance of KeyValueStore, a function f, and map m, and
+  updates the store in a manner similar to `clojure.core/swap!`."
+  [^KeyValueStore store f m]
+  (let [ks (keys (f {} m))
+        prev (reduce (fn [p k]
+                       (assoc p k (.get store k)))
+                     {}
+                     ks)
+        next (f prev m)]
+    (doall (map (fn [[k v]] (.put store k v)) next))
+    next))
+
+(defn add-state-store!
+  [builder]
+  "Takes a builder and adds a state store."
+  (doto ^StreamsBuilder (j/streams-builder* builder)
+    (.addStateStore (Stores/keyValueStoreBuilder
+                     (Stores/persistentKeyValueStore "transducer")
+                     (js/edn-serde)
+                     (js/edn-serde))))
+  builder)
+
+(defn transformer
+  "Takes a transducer and creates an instance of
+  org.apache.kafka.streams.kstream.Transformer with overrides for
+  init, transform, and close."
+  [xf]
+  (let [ctx (atom nil)]
+    (reify
+      Transformer
+      (init [_ context]
+        (reset! ctx context))
+      (transform [_ k v]
+        (let [^KeyValueStore store (.getStateStore @ctx "transducer")
+              v (first (into [] (xf store) [[k v]]))]
+          (KeyValue/pair k v)))
```

Comment:
This transform does not mess with the key - for clarity we should consider providing a `transformer` (which can affect the key) and a `value-transformer` which won't (i.e calls to the underlying `transform` and `transformValues` on the streams DSL).

Similarly for `transduce-kstream` vs `transduce-kstream-values`

This separation is useful when reading the high level code as it telegraphs to the reader whether a repartition may be occurring (a-la `map` vs `mapValues`)

Reply:
Each KStream::transform is always followed by a KStream::flatMap. The final key or keys are obtained from the resulting value of KStream::transform. We don't care about the resulting key. 

Suppose the input to KStream::transform is `[input-key input-value]`. The output might be `[input-key [[k1 v1] [k2 v2]]`. KStream::flatMap discards the key from the previous step and publishes two records with keys k1 and k2.

Reply:
that makes sense. but might not be clear to someone - if we decide not to replace it with `.forward` and `.commit`, i think this dependency should be documented. or `transformer` becomes a private function so it is clear `transduce-kstream` is the API to use